kvdb/postgres/readwrite_bucket.go

```diff
@@ -0,0 +1,399 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+)
+
+// readWriteBucket stores the bucket id and the buckets transaction.
+type readWriteBucket struct {
+	// id is used to identify the bucket. If id is null, it refers to the
+	// root bucket.
+	id *int64
+
+	// tx holds the parent transaction.
+	tx *readWriteTx
+
+	table string
+}
+
+// newReadWriteBucket creates a new rw bucket with the passed transaction
+// and bucket id.
+func newReadWriteBucket(tx *readWriteTx, table string, id *int64) *readWriteBucket {
+	return &readWriteBucket{
+		id:    id,
+		tx:    tx,
+		table: table,
+	}
+}
+
+// NestedReadBucket retrieves a nested read bucket with the given key.
+// Returns nil if the bucket does not exist.
+func (b *readWriteBucket) NestedReadBucket(key []byte) walletdb.ReadBucket {
+	return b.NestedReadWriteBucket(key)
+}
+
+func parentSelector(id *int64) string {
+	if id == nil {
+		return "parent_id IS NULL"
+	}
+	return fmt.Sprintf("parent_id=%v", *id)
+}
+
+// ForEach invokes the passed function with every key/value pair in
+// the bucket. This includes nested buckets, in which case the value
+// is nil, but it does not include the key/value pairs within those
+// nested buckets.
+func (b *readWriteBucket) ForEach(cb func(k, v []byte) error) error {
+	rows, err := b.tx.tx.QueryContext(
+		context.TODO(),
+		"SELECT key,value FROM "+b.table+" WHERE "+parentSelector(b.id)+" ORDER BY key",
+	)
+	if err != nil {
+		return err
+	}
+
+	type kv struct {
+		k, v []byte
+	}
+
+	// Read all data in memory, to unblock the connection (nested queries).
+	//
+	// TODO: Convert to paginated reads to prevent OOM on large tables.
+	var kvs []kv
+	for rows.Next() {
+		var (
+			key   []byte
+			value []byte
+		)
+		err := rows.Scan(&key, &value)
+		if err != nil {
+			rows.Close()
+			return err
+		}
+
+		kvs = append(kvs, kv{k: key, v: value})
```

Comment:
Good point. This is temporary code that I hadn't improved yet. Unfortunately it is not that easy. The reason for reading all in memory is that in some places, ForEach loops are nested. With sql, it isn't allowed to execute a `SELECT` while another `SELECT` result is still being streamed back. Only one query per transactions.

Reading all in memory is definitely a bad idea with OOM risk as you say. The question is how to do it otherwise :thinking: 

For read txes, I think it is safe to open a new transaction. But for writes it should indeed read back the pending changes. Not sure if that actually happens anywhere, but it is a concern nonetheless.

How would you go about this?

Reply:
The safe way is probably to use a cursor and fetch the rows one by one. Will need to check the performance impact of that.

Reply:
Converted to use a cursor.